lang/java/avro/src/main/java/org/apache/avro/generic/GenericDatumWriter.java

```diff
@@ -229,8 +230,17 @@ protected AvroTypeException addAvroTypeMsg(AvroTypeException e, String s) {
    * representations.
    */
   protected void writeRecord(Schema schema, Object datum, Encoder out) throws IOException {
-    Object state = data.getRecordState(datum, schema);
-    for (Field f : schema.getFields()) {
+    final Schema realSchema;
+    if (datum instanceof GenericRecord && schema.hasChild()) {
+      realSchema = ((GenericRecord) datum).getSchema();
+    } else {
+      realSchema = schema;
+    }
+    if (!Objects.equals(realSchema, schema)) {
+      out.writeExtends(realSchema.getIndex());
+    }
```

Comment:
This creates two possible binary representations of the same (parent) schema: one for writing a parent schema, and a different one for child schemata.

How can the reader distinguish between the two?

Reply:
Don't get this comment : 
- schema variable represent the schema declaration (For a field or an array of)
- realSchema variable represent the effective schema of the record (datum).
if it's the same, no pb, we stay in classic situation, but if it's not the same, we mark it as extended type.

Reply:
Fixed